tests/system/src/test/java/test/javafx/scene/control/behavior/BehaviorRobotTestBase.java

```diff
@@ -0,0 +1,346 @@
+     * For each item:
+     * <ul>
+     * <li> if a String, emulates KEY_PRESSED + KEY_TYPED + KEY_RELEASED events for each character
+     * <li> if a KeyCode, emulates KEY_PRESSED + KEY_RELEASED events for that KeyCode
+     * <li> if a Runnable, runs it
+     * </ul>
+     * @param items the sequence of KeyCodes/Runnables/String
+     */
+    protected void execute(Object ... items) {
+        for (Object x : items) {
+            if (x instanceof Runnable r) {
+                Util.runAndWait(() -> {
+                    r.run();
+                });
+            } else if (x instanceof KeyCode k) {
+                Util.runAndWait(() -> {
+                    robot.keyPress(k);
+                    robot.keyRelease(k);
+                });
+            } else if (x instanceof String s) {
+                for (int i = 0; i < s.length(); i++) {
+                    char c = s.charAt(i);
+                    KeyCode k = getKeyCodeForChar(c);
+                    Util.runAndWait(() -> {
+                        robot.keyPress(k);
+                        robot.keyRelease(k);
+                    });
+                }
+            }
+            step++;
+        }
+    }
+
+    /**
+     * Looks up a KeyCode for the given character.
+     * @param c character
+     * @return KeyCode
+     */
+    // TODO duplicated code from KeyEventFirer - we don't have a common place for general purpose test utilities
+    private static KeyCode getKeyCodeForChar(char c) {
+        if (keyCodes == null) {
+            keyCodes = createKeyCodes(
+                " ", KeyCode.SPACE,
+                "\t", KeyCode.TAB,
+                "\n", KeyCode.ENTER,
+                "a", KeyCode.A,
+                "b", KeyCode.B,
+                "c", KeyCode.C,
+                "d", KeyCode.D,
+                "e", KeyCode.E,
+                "f", KeyCode.F,
+                "g", KeyCode.G,
+                "h", KeyCode.H,
+                "i", KeyCode.I,
+                "j", KeyCode.J,
+                "k", KeyCode.K,
+                "l", KeyCode.L,
+                "m", KeyCode.M,
+                "n", KeyCode.N,
+                "o", KeyCode.O,
+                "p", KeyCode.P,
+                "q", KeyCode.Q,
+                "r", KeyCode.R,
+                "s", KeyCode.S,
+                "t", KeyCode.T,
+                "u", KeyCode.U,
+                "v", KeyCode.V,
+                "w", KeyCode.W,
+                "x", KeyCode.X,
+                "y", KeyCode.Y,
+                "z", KeyCode.Z,
+                "0", KeyCode.DIGIT0,
+                "1", KeyCode.DIGIT1,
+                "2", KeyCode.DIGIT2,
+                "3", KeyCode.DIGIT3,
+                "4", KeyCode.DIGIT4,
+                "5", KeyCode.DIGIT5,
+                "6", KeyCode.DIGIT6,
+                "7", KeyCode.DIGIT7,
+                "8", KeyCode.DIGIT8,
+                "9", KeyCode.DIGIT9,
+                ".", KeyCode.PERIOD,
+                ",", KeyCode.COMMA
```

Comment:
On French layouts the DIGIT codes don't actually produce digits (you have to hold down Shift) and there's no KeyCode.PERIOD (also a shifted character). You don't seem to be using those characters but you might want to remove them from this table so no one is tempted to use them in the future.

Reply:
Interesting.  

It is my understanding though that for the purposes of this test it should work as expected.  I've added `TextAreaBehaviorRobotTest.testTyping()` - could you try running it with your setup please?

```
    @Test
    public void testTyping() throws Exception {
        execute(
            //addKeyListener(),
            "0123456789.,abracadabra",
            checkText("0123456789.,abracadabra", 23)
        );
    }
```

Reply:
I ran the TextAreaBehaviorRobotTest on macOS 13.5.2. Test passed with my keyboard set to U.S. English and failed when set to French.

```
TextAreaBehaviorRobotTest > testTyping() FAILED
    org.opentest4j.AssertionFailedError: in step 2 ==> expected: <0123456789.,abracadabra> but was: <à&é"'(§è!ç,abracadabra>
```

Reply:
Merci, Monsieur Fox!

We seem to have quite a few tickets in this area (who created the first one in the list?):

* [JDK-8278924](https://bugs.openjdk.org/browse/JDK-8278924) [Linux] Robot key test can fail if multiple keyboard layouts are installed
* [JDK-8022079](https://bugs.openjdk.org/browse/JDK-8022079) [macosx] swing shortcuts are counterintuitive with non-english keyboard layouts
* [JDK-8087915](https://bugs.openjdk.org/browse/JDK-8087915) Mac: accelerator doesn't take into account azerty keyboard layout
* [JDK-4760902](https://bugs.openjdk.org/browse/JDK-4760902) java.awt.Robot can't generate % character on FRENCH keyboard

Not only we get key events for wrong characters, but also shortcut(A) generates `character = q, text = , code = UNDEFINED, metaDown, shortcutDown]` which closes the test app :-(

I have no idea how to fix this test...  We could try to introduce a translation layer to emit the right sequences (for each keyboard layout), or, alternatively, we could skip the tests altogether for non-US keyboard, except ... there seems to be no easy way to determine the keyboard layout?  (There are some suggestions involving InputContext or JNI on stackoverflow).  We could also try to skip the test if Locale.getDefault is other than Locale.US, not sure if that would solve the issue.

What do you think?

Reply:
The problem you're running into here isn't due to any bug in JavaFX. The way KeyCodes identify keys works well for defining accelerators but not for using a Robot to generate a specific printable character. As an extreme example, if you're using a Greek keyboard there will be a key that has KeyCode.Z assigned to it so you can use Shortcut+Z to invoke Undo. But that key will generate a Greek letter when pressed by a human or a Robot.

You can probably assume that anyone running these tests is using a Latin/Roman keyboard that can generate the lower-case letters a-z so one approach is to just stick to those characters in the typing tests. The other is to try to identify the layout using a Robot. For example, send KeyCode.DIGIT1 and see if that generates a '1'. If not you're probably on a French layout.

Punctation and symbols are hopeless, there's just too much variation in the keyboard layouts.

> Not only we get key events for wrong characters, but also shortcut(A) generates `character = q, text = , code = UNDEFINED, metaDown, shortcutDown]` which closes the test app :-(

Could you provide more details (platform, keyboard, keyboard layout, etc.)? On the Mac PR #425 cleaned all this up in jfx21.

Reply:
I am running this on macOS 13.5.1 with several keyboard layouts enabled.

A test that does keyPress(KecCode.A) followed by a keyRelease(KeyCode.A) generates the following sequence of events:

With the US layout:

```
KeyEvent [source = TextArea@7da926e[styleClass=text-input text-area], target = TextArea@7da926e[styleClass=text-input text-area], eventType = KEY_PRESSED, consumed = false, character = , text = a, code = A]
KeyEvent [source = TextArea@7da926e[styleClass=text-input text-area], target = TextArea@7da926e[styleClass=text-input text-area], eventType = KEY_TYPED, consumed = false, character = a, text = , code = UNDEFINED]
KeyEvent [source = TextArea@7da926e[styleClass=text-input text-area], target = TextArea@7da926e[styleClass=text-input text-area], eventType = KEY_RELEASED, consumed = false, character = , text = a, code = A]
```

With the French layout:

```
KeyEvent [source = TextArea@75ffb50a[styleClass=text-input text-area], target = TextArea@75ffb50a[styleClass=text-input text-area], eventType = KEY_PRESSED, consumed = false, character = , text = q, code = A]
KeyEvent [source = TextArea@75ffb50a[styleClass=text-input text-area], target = TextArea@75ffb50a[styleClass=text-input text-area], eventType = KEY_TYPED, consumed = false, character = q, text = , code = UNDEFINED]
KeyEvent [source = TextArea@75ffb50a[styleClass=text-input text-area], target = TextArea@75ffb50a[styleClass=text-input text-area], eventType = KEY_RELEASED, consumed = false, character = , text = q, code = A]
```

And with a cyrillic layout it generates whatever cyrillic letter is mapped to the key.  So, basically, KeyCode is more like a scan code, which means there is no reliable way to test key bindings using Robot, unless we limit ourselves (and the test) to the US layout.  Or is there another way?

<img width="612" alt="Screenshot 2023-09-18 at 13 19 51" src="https://github.com/openjdk/jfx/assets/107069028/8bd8745c-8973-449f-bf17-0473f5c0e0b6">

Reply:
This is the behavior we saw on the Mac before PR #425 was integrated. As of JavaFX 21 keyPress(KeyCode.A) should generate an 'a' and never a 'q' on all Latin layouts and all platforms (on a Cyrillic layout you'll still get a Cyrillic character).

I can't reproduce this with the test app (tests/manual/events/KeyboardTest.java). Is there a test in this PR I should be trying?

Reply:
You could try this code (using this PR's code):

```
public class KeyboardLayoutTest extends TextInputBehaviorRobotTest<TextArea> {

    public KeyboardLayoutTest() {
        super(new TextArea());
    }

    @BeforeEach
    @Override
    public void beforeEach() {
        super.beforeEach();
        control.setWrapText(true);
    }

    @Test
    public void testTyping() throws Exception {
        KeyCode[] codes = {
            KeyCode.A
        };

        execute(
            exe(() -> {
                control.addEventFilter(KeyEvent.ANY, (ev) -> {
                    System.out.println(ev);
                    ev.consume();
                });
            })
        );

        for(KeyCode cd: codes) {
            Util.runAndWait(() -> {
                robot.keyPress(cd);
                robot.keyRelease(cd);
            });
        }
    }
}
```

Reply:
I tried your test code and couldn't reproduce the problem. Both U.S. English and French layouts gave me the expected result (an 'a' character). I also tried a Russian keyboard just to verify that it would generate a Cyrillic character which it did. I'm running the tests against JavaFX built using the sources in your pull request.

French

```
KeyEvent [source = TextArea@7237703[styleClass=text-input text-area], target = TextArea@7237703[styleClass=text-input text-area], eventType = KEY_PRESSED, consumed = false, character = ?, text = a, code = A]
KeyEvent [source = TextArea@7237703[styleClass=text-input text-area], target = TextArea@7237703[styleClass=text-input text-area], eventType = KEY_TYPED, consumed = false, character = a, text = , code = UNDEFINED]
KeyEvent [source = TextArea@7237703[styleClass=text-input text-area], target = TextArea@7237703[styleClass=text-input text-area], eventType = KEY_RELEASED, consumed = false, character = ?, text = a, code = A]
```

U.S. English

```
KeyEvent [source = TextArea@73459ba5[styleClass=text-input text-area], target = TextArea@73459ba5[styleClass=text-input text-area], eventType = KEY_PRESSED, consumed = false, character = ?, text = a, code = A]
KeyEvent [source = TextArea@73459ba5[styleClass=text-input text-area], target = TextArea@73459ba5[styleClass=text-input text-area], eventType = KEY_TYPED, consumed = false, character = a, text = , code = UNDEFINED]
KeyEvent [source = TextArea@73459ba5[styleClass=text-input text-area], target = TextArea@73459ba5[styleClass=text-input text-area], eventType = KEY_RELEASED, consumed = false, character = ?, text = a, code = A]
```

Reply:
Thank you for your help!  I can confirm that the latest code does indeed generate the right key event.

I suspect there was something in my setup that got corrupted - I had to install the latest eclipse, found out it's pretty much unusable due to https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/309 , had to revert back to 2023-06 just to discover that my macOS happily suggested that 'your application is damaged and can't be opened' and do I want to move it to Trash?  So after some xattr'ing, restoring proxy settings, etc., etc. I am back online and the issue is gone.

I did remove the offending characters and modified the tests, so thank you once again - your help was invaluable.